#2024/1 - 1118003 - 1 - TÓPICOS ESPECIAIS EM COMPUTAÇÃO II
#Processamento de Linguagem Natural - PLN
#Guilherme Dallmann Lima

#Biblioteca

In [ ]:
from collections import Counter
import math

#Cálculo da Entropia

In [ ]:
def calculate_entropy(word):

    char_count = Counter(word)

    total_chars = sum(char_count.values())

    char_probabilities = {char: count / total_chars for char, count in char_count.items()}

    entropy = sum(-prob * math.log2(prob) for prob in char_probabilities.values())

    return entropy

In [ ]:
words = ["ARARAQUARA", "UFPEL", "PELOTAS"]

for word in words:
    entropy = calculate_entropy(word)
    print(f"A entropia da palavra '{word}' é: {entropy}")

A entropia da palavra 'ARARAQUARA' é: 1.6854752972273346
A entropia da palavra 'UFPEL' é: 2.321928094887362
A entropia da palavra 'PELOTAS' é: 2.807354922057604


#Explicação

ARARAQUARA: A entropia é a mais baixa (1.685) devido à alta repetição das letras "A" e "R", resultando em menor diversidade de caracteres.
UFPEL: Tem uma entropia maior (2.321) porque todas as letras são únicas, o que aumenta a diversidade e, consequentemente, a entropia.
PELOTAS: A entropia mais alta (2.807) indica uma distribuição bastante uniforme de letras, com poucas repetições, tornando a palavra mais imprevisível e complexa.